CactbotOverlay/FFXIVProcess.cs

```diff
@@ -2,1261 +2,1559 @@
-      data.id = BitConverter.ToUInt32(bytes, kEntityStructureOffsetId);
-      data.type = (EntityType)bytes[kEntityStructureOffsetType];
-      data.distance = bytes[kEntityStructureOffsetDistance];
-      data.pos_x = BitConverter.ToSingle(bytes, kEntityStructureOffsetPos);
-      data.pos_z = BitConverter.ToSingle(bytes, kEntityStructureOffsetPos + 4);
-      data.pos_y = BitConverter.ToSingle(bytes, kEntityStructureOffsetPos + 8);
-
-      if (data.type == EntityType.PC || data.type == EntityType.Monster) {
-        data.job = (EntityJob)bytes[kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetJob];
-
-        data.hp = BitConverter.ToInt32(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetHpMpTp);
-        data.max_hp = BitConverter.ToInt32(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetHpMpTp + 4);
-        data.mp = BitConverter.ToInt32(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetHpMpTp + 8);
-        data.max_mp = BitConverter.ToInt32(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetHpMpTp + 12);
-        var shieldPercentage = BitConverter.ToInt16(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetShieldPercentage);
-        data.shield_value = shieldPercentage * data.max_hp / 100;
-
-        if (IsGatherer(data.job)) {
-          data.gp = BitConverter.ToInt16(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetGpCp);
-          data.max_gp = BitConverter.ToInt16(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetGpCp + 2);
-        } else {
-          data.gp = 0;
-          data.max_gp = 0;
-        }
-        if (IsCrafter(data.job)) {
-          data.cp = BitConverter.ToInt16(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetGpCp + 4);
-          data.max_cp = BitConverter.ToInt16(bytes, kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetGpCp + 6);
-        } else {
-          data.cp = 0;
-          data.max_cp = 0;
-        }
-        data.level = bytes[kEntityStructureOffsetCharacterDetails + kEntityStructureOffsetLevel];
-
-        byte[] job_bytes = GetJobSpecificData();
-        data.debug_job = "";
-
-        if (job_bytes != null) {
-          // Start at 8 to skip past the initial pointer.
-          for (var i = 8; i < job_bytes.Length; ++i) {
-            if (data.debug_job != "")
-              data.debug_job += " ";
-            data.debug_job += string.Format("{0:x2}", job_bytes[i]);
-          }
-        }
+
+      private byte[] GetJobSpecificData() {
+         if (!HasProcess() || job_data_outer_addr_ == IntPtr.Zero)
+            return null;
+
+         IntPtr job_inner_ptr = ReadIntPtr(job_data_outer_addr_);
+         if (job_inner_ptr == IntPtr.Zero) {
+            // The pointer can be null when not logged in.
+            return null;
+         }
+
+         return Read8(job_inner_ptr, kJobDataInnerStructSize);
```

Comment:
Given that everything starts at 0x8, would you mind adding that offset in here, so that all the job memory classes start at 0x0?